### ISSUE TITLE
Stack overflow on large strings 

### ISSUE THREAD

**Issue by JoshuaGross** [orlandohill/waxeye#23]
I'm getting stack overflows for fairly simple patterns matched on large-ish strings. Some of my users have reported seeing this issue in Java; I've been testing in JavaScript and it's very easy to reproduce.

Grammar:

```
Program <= +.
```

Generate parser:
```
> waxeye -g javascript ./ grammar.waxeye
```

test.js:
```
var parser = require('./parser.js');
var assert = require('assert');
var parserInst = new parser.Parser();


// String.prototype.repeat polyfill from mozdev
if (!String.prototype.repeat) {
  String.prototype.repeat = function(count) {
    'use strict';
    if (this == null) {
      throw new TypeError('can\'t convert ' + this + ' to object');
    }
    var str = '' + this;
    count = +count;
    if (count != count) {
      count = 0;
    }
    if (count < 0) {
      throw new RangeError('repeat count must be non-negative');
    }
    if (count == Infinity) {
      throw new RangeError('repeat count must be less than infinity');
    }
    count = Math.floor(count);
    if (str.length == 0 || count == 0) {
      return '';
    }
    // Ensuring count is a 31-bit integer allows us to heavily optimize the
    // main part. But anyway, most current (August 2014) browsers can't handle
    // strings 1 << 28 chars or longer, so:
    if (str.length * count >= 1 << 28) {
      throw new RangeError('repeat count must not overflow maximum string size');
    }
    var rpt = '';
    for (;;) {
      if ((count & 1) == 1) {
        rpt += str;
      }
      count >>>= 1;
      if (count == 0) {
        break;
      }
      str += str;
    }
    return rpt;
  }
}

parseStrOfLength(3950);

function parseStrOfLength (len) {
	console.log(len);
	parserInst.parse('"' + 'a'.repeat(len) + '"')
}
```

Run test:
```
> node test.js
3950
/Users/jgross/Desktop/Projects/StackOverflowExample/node_modules/waxeye/waxeye.js:177
  InnerParser.prototype.matchEdges = function(edges, index) {
                                             ^
RangeError: Maximum call stack size exceeded
    at InnerParser.matchEdges (/Users/jgross/Desktop/Projects/StackOverflowExample/node_modules/waxeye/waxeye.js:177:46)
    at InnerParser.matchState (/Users/jgross/Desktop/Projects/StackOverflowExample/node_modules/waxeye/waxeye.js:174:16)
    at InnerParser.matchEdge (/Users/jgross/Desktop/Projects/StackOverflowExample/node_modules/waxeye/waxeye.js:195:22)
    at InnerParser.matchEdges (/Users/jgross/Desktop/Projects/StackOverflowExample/node_modules/waxeye/waxeye.js:182:18)
    at InnerParser.matchState (/Users/jgross/Desktop/Projects/StackOverflowExample/node_modules/waxeye/waxeye.js:174:16)
    at InnerParser.matchEdge (/Users/jgross/Desktop/Projects/StackOverflowExample/node_modules/waxeye/waxeye.js:195:22)
    at InnerParser.matchEdges (/Users/jgross/Desktop/Projects/StackOverflowExample/node_modules/waxeye/waxeye.js:182:18)
    at InnerParser.matchState (/Users/jgross/Desktop/Projects/StackOverflowExample/node_modules/waxeye/waxeye.js:174:16)
    at InnerParser.matchEdge (/Users/jgross/Desktop/Projects/StackOverflowExample/node_modules/waxeye/waxeye.js:195:22)
    at InnerParser.matchEdges (/Users/jgross/Desktop/Projects/StackOverflowExample/node_modules/waxeye/waxeye.js:182:18)
```

Not really sure how to simplify this sort of matcher or restructure it to prevent crashing. Any ideas?

**Comment by orlandohill** [comment]
Perhaps moving the repetition into recursive non-terminals would help
somewhat.

I've been away from my laptop for the last three and a half months. Will be
back tomorrow, so development can start again then.

**Comment by JoshuaGross** [comment]
So if I use this grammar:

```
Program <= *Text
Text <= .
```

Then it craps out around 7100 characters instead of 3950.

And this fails at 1750:

```
Program <= Text
Text <= . Text | .
```

Am I missing something / am I interpreting recursive NT incorrectly?

**Comment by JoshuaGross** [comment]
Also, welcome back!

**Comment by orlandohill** [comment]
The second example was what I had in mind. Doesn't look like it helped that
much.

Sorry that I can't offer a more useful solution just yet.

**Comment by JoshuaGross** [comment]
So after some investigation it looks like this recursion cycle is the culprit: `matchAutomaton -> matchState -> matchEdges -> matchEdge -> matchAutomaton -> ...`. 
Perhaps this was obvious to you. There's no obvious refactor that's jumping out at me so I'll keep investigating.

**Comment by JoshuaGross** [comment]
I'm playing around with moving everything to a stack-based architecture in the JavaScript code. Let me know if I'm duplicating your efforts. I don't know the code well so it's gonna take some time.

**Comment by orlandohill** [comment]
I've already designed a new version that separates the parsing stack from
the implementation language's runtime stack.

You can see how it is derived here.
https://github.com/orlandohill/waxeye/blob/master/src/sml/waxeye.sml

To better understand this version, I would recommend the following papers:

* Three Steps for the CPS Transformation (1991) by Olivier Danvy

* Section 3 of Defunctionalization at Work (2001) by Olivier Danvy, Lasse
R. Nielsen

* Section 2.1 of A Functional Correspondence between Evaluators and
Abstract Machines (2003) by Mads Sig Ager, Dariusz Biernacki, Olivier
Danvy, Jan Midtgaard

**Comment by JoshuaGross** [comment]
Were you planning to implement this in any of the (other) host languages? I'm working on the JavaScript implementation but I'm pretty strapped for time, if you'd be able to tackle C and/or Java that would be greatly appreciated (JS, C, and Java are the three we're using at Yahoo).

Thanks for the link and papers, checking them out now!

**Comment by JoshuaGross** [comment]
Are Match0 ... MatchN all evolutions of the same function? Each would be used standalone?

**Comment by orlandohill** [comment]
I am going to implement the new parsing machine design for all the
supported host languages.

Yes, Match0-4 are all implementations of the same set of functions. Each
language only needs one. Match0 for Scheme/Racket because it can handle the
recursion, Match4 for the others.

**Comment by JoshuaGross** [comment]
Great, I'm maybe 20% done with the JavaScript version. I may be able to handle the Java version myself but would definitely appreciate help with the C implementation.

**Comment by JoshuaGross** [comment]
I'm done with the JavaScript version and will submit a PR once I have open-source approval from Yahoo. One thing that's made my life a lot easier is using a test-fixtures file, where all the tests are encoded as JSON and processed by a JavaScript file. I suggest we do this for all the languages instead of rewriting tests in each language.

**Comment by orlandohill** [comment]
Glad to hear!

Yes, I was thinking of having a DSL for the tests, and generating source
code to construct the test data. Putting them in a JSON file seems a good
solution.

**Comment by JoshuaGross** [comment]
The Javascript engine and its tests are all done (again, I'll submit a PR pending open-source approval) but I'm having trouble with the `javascript.scm` code generator, probably in particular because I understand the new parsing engine much better than the previous one. Anyway, if you'd be willing to update the code generation then I can start on the Java implementation of the new engine.

**Comment by orlandohill** [comment]
Yes, I'm happy to update the code generation.

The generation process should get a lot simpler, because the data
structures used by the new parsing engine closely mirror those obtained by
parsing a grammar file.

**Comment by JoshuaGross** [comment]
Yeah, that's what I figured. I should also specify that I've never written Scheme in my life so I also felt like my time was better spent on the Java side for now. Thanks!

**Comment by adabru** [comment]
Hello, thank you very much for your public work. I like waxeye very much to create an individually featured markup-language. I also sometimes run in the problem of 'too much recursion' and would be glad about an updated generator for the new algorithm. Is work going on on waxeye or will there be a longer break?

**Comment by newsoft** [comment]
I am hitting the same issue with C code. Here are repro instructions; the program will segfault due to stack exhaustion: https://github.com/newsoft/waxeye/commit/194131fa794b4142e3ce8be2e86d53f13c68204e

**Comment by orlandohill** [comment]
Hi Nicolas,

Sorry about that. Thanks for taking the time to report this.

This is a known problem with the old design of the parsing machine, for all
implementation languages except Scheme.

I should more clearly mark that the project is "under construction" and
looking for volunteers.

A new parsing machine has been designed that fixes this problem.

Converting to the new design should only take a weekend or two per
implementation language. Unfortunately, I'm not able to do it myself at the
moment.

I'm not sure who, if anyone, is currently working on the C version. You
might like to check with @glebm @ddrone @jishi9
